File: frontend/src/types/reasoning.ts
export type ReasoningStrategy = 'association' | 'hypothetico' | 'constraints' | 'arguments';

export interface ReasoningStep {
    step_number: number;
    description: string;
    evidence?: string[];
    supports?: string[];  // option IDs supported (e.g., 'A', 'B')
    opposes?: string[];   // option IDs opposed
}

export interface ReasoningResult {
    strategy: ReasoningStrategy;
    question_id: string;
    steps: ReasoningStep[];
    conclusion: string;
    confidence: number;
    correct_option: string;
}

// Association reasoning - uses knowledge graph connections
export interface AssociationReasoning extends ReasoningResult {
    strategy: 'association';
    key_findings: string[];
    linked_conditions: Array<{
        condition: string;
        strength: number;
        matching_findings: string[];
    }>;
}

// Hypothetico-deductive reasoning - hypothesis testing approach
export interface HypotheticoReasoning extends ReasoningResult {
    strategy: 'hypothetico';
    hypotheses: Array<{
        option: string;
        hypothesis: string;
        predictions: string[];
        verified: boolean;
        falsified: boolean;
    }>;
}

// Constraint satisfaction - elimination approach
export interface ConstraintReasoning extends ReasoningResult {
    strategy: 'constraints';
    constraints: Array<{
        finding: string;
        eliminates: string[];  // options eliminated by this finding
        reason: string;
    }>;
    remaining_options: string[];
}

// Argument-based reasoning - pros/cons for each option
export interface ArgumentReasoning extends ReasoningResult {
    strategy: 'arguments';
    arguments: Array<{
        option: string;
        pros: string[];
        cons: string[];
        net_score: number;
    }>;
}

export type AllReasoningResult = AssociationReasoning | HypotheticoReasoning | ConstraintReasoning | ArgumentReasoning;

File: backend/app/models/reasoning.py
from pydantic import BaseModel
from typing import Optional, Literal
from enum import Enum


class ReasoningStrategy(str, Enum):
    ASSOCIATION = "association"
    HYPOTHETICO = "hypothetico"
    CONSTRAINTS = "constraints"
    ARGUMENTS = "arguments"


class ReasoningStep(BaseModel):
    step_number: int
    description: str
    evidence: Optional[list[str]] = None
    supports: Optional[list[str]] = None
    opposes: Optional[list[str]] = None


class LinkedCondition(BaseModel):
    condition: str
    strength: float
    matching_findings: list[str]


class AssociationReasoning(BaseModel):
    strategy: Literal["association"]
    question_id: str
    steps: list[ReasoningStep]
    conclusion: str
    confidence: float
    correct_option: str
    key_findings: list[str]
    linked_conditions: list[LinkedCondition]


class Hypothesis(BaseModel):
    option: str
    hypothesis: str
    predictions: list[str]
    verified: bool
    falsified: bool


class HypotheticoReasoning(BaseModel):
    strategy: Literal["hypothetico"]
    question_id: str
    steps: list[ReasoningStep]
    conclusion: str
    confidence: float
    correct_option: str
    hypotheses: list[Hypothesis]


class Constraint(BaseModel):
    finding: str
    eliminates: list[str]
    reason: str


class ConstraintReasoning(BaseModel):
    strategy: Literal["constraints"]
    question_id: str
    steps: list[ReasoningStep]
    conclusion: str
    confidence: float
    correct_option: str
    constraints: list[Constraint]
    remaining_options: list[str]


class Argument(BaseModel):
    option: str
    pros: list[str]
    cons: list[str]
    net_score: int


class ArgumentReasoning(BaseModel):
    strategy: Literal["arguments"]
    question_id: str
    steps: list[ReasoningStep]
    conclusion: str
    confidence: float
    correct_option: str
    arguments: list[Argument]
